homeassistant/components/arcam_fmj/media_player.py

```diff
@@ -0,0 +1,355 @@
+    data = hass.data[DOMAIN_DATA_ENTRIES][config_entry.entry_id]
+    client = data["client"]
+    config = data["config"]
+
+    host = config[CONF_HOST]
+    port = config[CONF_PORT]
+
+    async_add_entities(
+        [
+            ArcamFmj(
+                State(client, zone),
+                zone_config.get(
+                    CONF_NAME,
+                    "{} ({}:{}) - {}".format(DEFAULT_NAME, host, port, zone),
+                ),
+                zone_config.get(SERVICE_TURN_ON),
+            )
+            for zone, zone_config in config[CONF_ZONE].items()
+        ]
+    )
+
+    return True
+
+
+class ArcamFmj(MediaPlayerDevice):
+    """Representation of a media device."""
+
+    def __init__(self, state: State, name: str, turn_on: Optional[ConfigType]):
+        """Initialize device."""
+        super().__init__()
+        self._state = state
+        self._name = name
+        self._turn_on = turn_on
+        self._support = (
+            SUPPORT_SELECT_SOURCE
+            | SUPPORT_VOLUME_SET
+            | SUPPORT_VOLUME_MUTE
+            | SUPPORT_VOLUME_STEP
+            | SUPPORT_TURN_OFF
+        )
+        if state.zn == 1:
+            self._support |= SUPPORT_SELECT_SOUND_MODE
+
+    def _get_2ch(self):
+        """Return if source is 2 channel or not."""
+        audio_format, _ = self._state.get_incoming_audio_format()
+        return bool(
+            audio_format
+            in (
+                IncomingAudioFormat.PCM,
+                IncomingAudioFormat.ANALOGUE_DIRECT,
+                None,
+            )
+        )
+
+    @property
+    def unique_id(self) -> Optional[str]:
+        """Return a unique ID."""
+        return "{}_{}_{}".format(
+            self._state.client.host, self._state.client.port, self._state.zn
+        )
+
+    @property
+    def device_info(self):
+        """Return a device description for device registry."""
+        return {
+            "identifiers": {
+                (DOMAIN, self._state.client.host, self._state.client.port)
+            },
+            "model": "FMJ",
+            "manufacturer": "Arcam",
+        }
+
+    @property
+    def should_poll(self) -> bool:
+        """No need to poll."""
+        return False
+
+    @property
+    def name(self):
+        """Return the name of the controlled device."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state of the device."""
+        if self._state.get_power():
+            return STATE_ON
+        return STATE_OFF
+
+    @property
+    def supported_features(self):
+        """Flag media player features that are supported."""
+        support = self._support
+        if self._state.get_power() is not None or self._turn_on:
+            support |= SUPPORT_TURN_ON
+        return support
+
+    async def async_added_to_hass(self):
+        """Once registed add listener for events."""
+        await self._state.start()
+
+        def _data(host):
```

Comment:
The callbacks must be annotated with `@callback` imported from core.py, since they should be run in the event loop.